### PR TITLE
Add file transfer capabilities

### PR DESCRIPTION
**Pull request by robogen**
This merge is to bring the updates generated in this fork to the main repo. To this point the following changes have been made.

1. Gruntfile.js modified to require the new updated packages.
2. Framework for single and multiple file upload established
3. Upload page designed and setup.
4. node.js file managment method changed to fix issue in backend processing. Change exists in Gruntfile.js, changing sort from ByType to ByComponent.
5. Top navigation bar modified to exist in single webpage and is referenced instead of being unique in each page.

uploadfile
uploadfolder

Other modifications still required.

Yet to be done:
- [x] Single file upload works fine currently, but folder upload needs modification. Currently loops back and runs 'make folder' and 'upload file' on same folder/file multiple times. Behind the scenes code modification required as well to fix asynchronous upload bug.

Will be done later:
1. 'Dropbox' feature needs further development, first step would be to fix current issues with folder upload feature
2. Stylization has a drastic difference between Mac/Linux and Windows. Not sure if something can be done about this.
3. Detecting file size of files uploading/downloading and report if too large for destination, or if file may be too large for destination.

### CODE REVIEW

**Comment by djw8605** [comment]
Lets not worry about cluster switching in this pull request.  Just focus on the upload / download of files and directories.

File: src/js/ConnectionService.js [hunk 1]
```diff
@@ -1,150 +1,371 @@
...
-        });
-        
-
-      });
-      
+	var connectionList = {crane: null,
+							tusker: null,
+							sandhills: null,
+							glidein: null};
+	$log.debug(connectionList);
```
Comment (djw8605) [inline src/js/ConnectionService.js]: It looks like in this commit, you are simply converting a couple spaces into a tab.  Please only use spaces.

File: src/js/app.js [hunk 2]
```diff
@@ -9,8 +12,39 @@ var app = angular.module('HccGoApp', ['HccGoApp.WelcomeCtrl', 'ngRoute', 'Connec
     }).when('/cluster/:clusterId', {
       templateUrl: 'html/clusterLanding.html',
       controller: 'clusterLandingCtrl'
-    }).otherwise({
+    }).when('/cluster/:clusterId/upload', {
+		templateUrl: 'html/clusterFileUpload.html',
+		controller: 'clusterUploadCtrl'
+	}).when('/cluster/:clusterId/download', {
```
Comment (djw8605) [inline src/js/app.js]: Another tabbing

File: src/js/ConnectionService.js (hunk 1 above)
Comment (robogen) [inline src/js/ConnectionService.js]: Simple enough. Not sure why the tabbing is so drastic here.

Comment (djw8605) [inline src/js/ConnectionService.js]: Ya, the tabbing is throughout.  Please go through your files and remove all tabs!

**Comment by djw8605** [comment]
Also, could you edit the pull request comment at the top so that the pictures are after the list of things that are done.  Currently, they are between number 3 and 4 on the list.

**Comment by robogen** [comment]
Made the modification to remove tab spaces.

**Comment by djw8605** [comment]
As far as I can tell, there's only 1 item left to finish this pull request:

> Single file upload works fine currently, but folder upload needs modification. Currently loops back and runs 'make folder' and 'upload file' on same folder/file multiple times. Behind the scenes code modification required as well to fix asynchronous upload bug.

**Comment by djw8605** [comment]
I tested the pull request.  Few comments:

- The 'upload' button on the main page doesn't match the style of the Submit Job
image
-We really need to figure out how to align the arrow better.
- Errors when I try to upload files:
image

**Comment by robogen** [comment]
Moved the controller for the nav bar to its own file, and fixed the 'listing' error. Not sure why that was there, it was a remnant of debugging from before. I am now getting an error where the function 'fastPut' becomes undefined when uploading a directory. It works just fine when uploading a single file, and both functions for uploading a single file and a directory refer to the same 'upload' function, so not sure why the context changes for it and it becomes undefined.

On the note of the upload button, what if we move it to the nav bar area? It's a rather large button, and takes up a lot of real estate on the screen when we have displays for data that is relevant to the user in the same spot.

**Comment by djw8605** [comment]
I don't see anything obvious on the fastPut becoming undefined.  Isn't fastput in the ConnectionService?  So nothing in the upload (higher level code) is calling it directly?

Could you upload an example screen shot of putting the upload files in the nav bar?

**Comment by robogen** [comment]
Yea, fastPut is a function of the SFTP service in ConnectionService, so nothing calls it directly.

Here's those screenshots, something of how it could look: 
file_nodrop
file_drop

**Comment by robogen** [comment]
Think I figured out the issue; the SFTP object can only be called so many times within a couple seconds, and there's a section of code that apparently calls the same section of code in ConnectionService multiple times very quickly. I'm working on using the Semaphore function to fix it. I'm running into a bug because of it, but the bug is in a section of code I wanted to rework anyway (makeDir function in ConnectionService) so it's a good opportunity for that.

Another thought: we could move the 'Submit' button to the navbar as well since it's just a page redirect, if you like the idea of putting the file transfer stuff in the navbar. Cleans up the page in clusterLanding, and allows control of the application from whichever page you are currently on.

**Comment by djw8605** [comment]
Lets save the navbar changes for a separate pull request.

It's odd that you can't call it fast.  I wonder if it's like the command interface, where you can only call it 1 at a time?

**Comment by robogen** [comment]
I think so. Using the semaphore technique I'm not experiencing the error message any more, and some files get uploaded with the directory method.

However, it just some files. The method getFiles() in clusterUploadCtrl is currently where I'm getting the bug experience. This is the section I originally thought had a tendency to create multiple entries for the same file, but it seems to just not be processing correctly. 

I have a test folder hierarchy that I'm testing against. The overall SFTPTest folder, with three folders under it called FolderOne, FolderTwo, and FolderThree. FolderOne contains no files, FolderTwo contains 'file.gif' and 'file.txt'. FolderThree contains 'file.doc' and 'file.xml'. They're all empty files.

Currently, when I try to upload this folder hierarchy, I don't get any outright error messages. However, the behaviour I see is that on the server FolderOne contains nothing (correct), FolderTwo contains 'file.txt' missing the 'file.gif', and FolderThree contains 'file.xml' and is missing 'file.doc'. 

Working on this right now.

**Comment by robogen** [comment]
I fixed the issue I was having, folder upload should work now. Mind giving it a test to verify?

Turns out, when the files were getting uploaded, they were uploading appropriately for how javascript works. I was having an issue because javascript doesn't maintain block state. I found the answer to my issue here: http://zef.me/2843/javascript-the-scope-pitfall

**Comment by djw8605** [comment]
Although that is an ok solution, couldn't you also just create the `name` variable inside the for loop anew each time?

**Comment by robogen** [comment]
That does seem to work, I'm a little surprised at that. Thought I had tried it before. I'll commit that change.

**Comment by robogen** [comment]
Lines removed

**Comment by djw8605** [comment]
I am still unable to successfully copy the source of HCCGo from my laptop to Crane.  It always `cannot read property of mkdir` error.

I wonder if we are trying to re-invent the wheel too much with this complicated.  I wonder if you could use / modify / copy some code from https://github.com/spmjs/node-scp2.  It does directory transfers and/or file transfers.  Additionally, you may be able to use it to transfer an file or directory without the user having to choose between the 2.

**Comment by robogen** [comment]
I confirmed the issue related too `cannot read property of mkdir`, too many folders get attempted to be created and the SFTP session limit gets hit (10).

I'm still looking into the scp2 package, which seems to solve a lot of the issues. We would loose the ability for an easy progress bar however, but that's about the only issue I can see.

Otherwise, seems NPM does have packages that support [tar](https://www.npmjs.com/package/tar) compression . A potential thing we could do is keep the current upload/download structure, but tar and compress the files and folders for upload and untar them when they arrive on the server. What do you think?

File: Gruntfile.js [hunk 3]
```diff
@@ -1,59 +1,45 @@
...
   grunt.initConfig({
-
-    nodewebkit: {
+    nwjs: {
        options: {
        	build_dir: './webkitbuilds',
-	platforms: ['osx','linux','win'],
-      	 mac: true,
-         version: '0.12.2'
+        platforms: ['linux'],
```
Comment (djw8605) [inline Gruntfile.js]: Please add osx, linux, and win.

File: src/package.json [hunk 4]
```diff
@@ -16,13 +18,18 @@
...
   },
   "webkit": {
     "plugin": true
   },
   "dependencies": {
     "ssh2": ">= 0.3.3",
-    "semaphore": ">= 1.0.2"
+    "semaphore": ">= 1.0.2",
+    "fs": ">= 0.0.2",
+    "scp2": ">= 0.2.2",
```
Comment (djw8605) [inline src/package.json]: Do we really need the scp2 package?

**Comment by djw8605** [comment]
For Feb. 26:
- [ ] Two column view of transferring files.
- [ ] Make progress bar work.
- [ ] Merge master
- [ ] Test on windows.


Future:
- [ ] Multiple SFTP connections.

**Comment by robogen** [comment]
This latest commit merges in the master at unlhcc

**Comment by robogen** [comment]
Graphically have the foundations for filemanagement setup, having a bit of trouble with AngularJS though. When the event to update the view is fired, the state that runs $scope.$apply in the view model finishes and exits before the view for local files can resolve completely, so it sees the array becomes empty but it doesn't update the view once content is added. However, causing a ng-click event on the remote view side will display the new content for local files (as 'ng-click' is effectively a wrapper for $scope.$apply). It's otherwise working, so temporarily going to leave this issue alone and finish with the full functionality of the display then return to it. I may need to start writing directives to get around the issue, so I'll start reading into how angularjs handles those.

**Comment by robogen** [comment]
Latest commit adds back the progress bar functionality, also adds a (X/N) counter to show how many files have been uploaded so far. Graphically it has some slight issues (progress bar not reaching its end for example) but programmatically works fine. This graphics issue is with the AngularJS processing model, I'll be looking into using directives to counteract it.

**Comment by robogen** [comment]
Commit fixes issue of local file view updating issue, so now without extra effort the user will see local files in the view. Next issue to work on will be to change the appearances of the upload/download buttons so they don't appear and disappear as much, and simply get disabled depending on what file is selected where. After that, I'll get the download function working.

**Comment by robogen** [comment]
99e31e9 changes behavior of upload/download buttons. I like this method a lot more, mind checking it out to see if it looks good? Otherwise my next goal is to start work on the actual Download functionality.

**Comment by robogen** [comment]
It should be working, it comes up for me. Though sometimes I get a little bit of weird progress bar behavior, I think due to the speed of some of the files uploading. Do you see the counter for files uploaded and total files to upload?

And I honestly hadn't considered adding the scroll bars to the file lists themselves, I'll see about adding those to it.

**Comment by robogen** [comment]
Scrolling on the divs works now, does the progress bar still not work appropriately for you?

**Comment by djw8605** [comment]
Scrolling works for up and down.  But the page still formats really weird horizontally.

Also, no, progress bar doesn't work:
image

And, I get an error when I try to transfer files:
image

**Comment by robogen** [comment]
Fixed the upload error, though not sure why it came up in the first place. And I'm really not sure why the progress bar and that div looks weird for you. Here is what I'm seeing: 

current

**Comment by djw8605** [comment]
Why are the values in the top right navbar breaking?

**Comment by robogen** [comment]
Oh, I never noticed. Turns out it's still referencing the navbar.html page I created from earlier. The NavCtrl is gone but its still pulling in the elements from the file. I'll fix that.

**Comment by robogen** [comment]
7e9c9e4 adds the Download function. Also is the first time I've implemented the async.queue method, in the readDir function. Seeing how it works now should help with the file upload/download parallelization later. I still have some of the graphical issues to work with, like with when file names are very long and extend the window, those are my next priority.

**Comment by robogen** [comment]
Closing was my bad, accidentally clicked the wrong button when commenting

**Comment by djw8605** [comment]
I tested the download.  The first time it worked.  The second, I get this error:
image

**Comment by robogen** [comment]
I'm not able to replicate the error. Did you just try to download the same folder twice, and the second time you got the error?

**Comment by djw8605** [comment]
Different folders.

-Derek

**Comment by robogen** [comment]
I'll keep playing around. I also found another error with downloading individual files, I'll try to fix those both up

**Comment by djw8605** [comment]
It always happens with the same directory.

image

**Comment by djw8605** [comment]
I believe the issue is transferring directories only 1 directory deep.  IE, a directory structure like:

```
$ find submit_condor/
submit_condor/
submit_condor/start_workers.sh
submit_condor/submit.sh
submit_condor/02_workers.conf
submit_condor/01_cached.conf
```

I've tested this theory against 2 different directories with this structure.

**Comment by robogen** [comment]
I think you were right, I saw a similar issue occurring with upload function. Seems the reason its occurring is because when there's subfolders available, the old code was able to create the overall folder to store everything because it got incorporated during the building of folders needing to be created. I fixed it by pushing the name of the origin folder into the array holding the listing of folders required to be generated. I performed a test with a simple folder structure I created locally, uploaded, then downloaded again to a new folder. It seems to be working now.

Still working on the single file download fix, shouldn't be too long on that though.

**Comment by robogen** [comment]
Latest commits add some slight modifications to the behaviour of the progress bar as files are being transferred. The most noticeable is after processing has finished, clicking on another file that's remote/local will hide the previous progress display, and a 'processing' message has been added that displays while the folder structure builds. This update will assist with the future modification of displaying the progress not for each file but for all files.

**Comment by robogen** [comment]
This commit modifies the progress bar functionality slightly, showing one for current file progress and one for total file progress. Currently only works for file uploads, though not perfectly. There's some math errors going on somewhere I need to find.

**Comment by robogen** [comment]
This commit is pretty similar to the previous one, but just adds the progress bar functionality to download process. 

I did have two questions for you, the first being do you want the 'current file' progress bar? In the case for the queued files, I would need to figure out a way to uniquely identify the callbacks and generate a progress bar. If we only wanted the 'total progress' bar, then that's pretty simple and I could potentially have the queued file upload/download procedure done relatively quickly. Which would you prefer?

The second question is I realized a simple method to make the progress bar a little bigger (simple CSS, not sure why I didn't try it before). Would you want the progress bars to be a bit bigger?

**Comment by djw8605** [comment]
I would like 1 and only 1 progress bar.  It should be for the entire transfer.

Can I get a screen shot of the progress bar now that it's bigger?  What about making it animated like:
https://fezvrasta.github.io/bootstrap-material-design/#progress-bar

**Comment by robogen** [comment]
bigger_progress
Something a little like that. I'll make the modification to reduce to just the one progress bar and adding the queue structure. And yes, its supposed to be animated like those bars.

**Comment by djw8605** [comment]
Could you stuff the progress bar under the "Download" and "Upload" box?

**Comment by robogen** [comment]
I hadn't thought of doing that, but I like the idea and see no reason why not. Sure, I'll give it a try.

**Comment by robogen** [comment]
moved_progress
This update adds in the capability of concurrently uploading and downloading files. The progress bar goes a little wonky as it gets conflicting information from callbacks. Not terrible, but I'll look into normalizing that a bit more.

This update also moves the progress bar to where you suggested, and I like this a lot more. Makes a lot more sense. 

The searching and folder creation portion of the download method is a little slow, need to figure out how to speed that up. Not sure if its a client/server thing or an algorithm thing, tackling both potentials.

**Comment by djw8605** [comment]
Great!  For the simultaneous downloads, could you create a ‘box’ around the progress information (bar and the total transferred) for each transfer.  Consider it a “queue” of transfers.  We don’t necessarily need to actually hold off on later transfers, they can all happen simultaneously.  But if we had a stack of transfers that would build up, and show each transfer’s progress.  I think that would look great!

In angular, this should translate to an angular list?

-Derek

**Comment by robogen** [comment]
I'm working on disk usage statistics, to warn users as the upload and download occurs if they will use up storage space on the local/remote systems. The package I'm looking to use for the local system is https://www.npmjs.com/package/diskusage and though it claims *nix support it doesn't really mention MacOS. Would you mind demoing it on your machine to confirm it works there? Also, is there a way to get Mac systems running on Anvil?

**Comment by djw8605** [comment]
Nope, Mac virtual machines do not work on Anvil (and really, Mac virtual machines are not really a thing).

**Comment by robogen** [comment]
Latest commit is a performance boost to the Download method. Previously, it took a very long time for the method to search through and create a folder hierarchy. Using the 'cargo' method in the async.js library instead of 'queue' I can recycle an SFTPSession to be used multiple times, saving time on both creation and clearing of old sessions. Previously took 10-12 minutes to search through folders on the remote host when downloading the HCCGo directory, with the new method it takes 2-4 minutes. I tried the same method of 'cargo' on the actual downloading of files method, I haven't timed it between the two versions but it seems to be as fast or faster by recycling the SFTPSession. Next I'll try to do the same to the upload method.

**Comment by robogen** [comment]
Latest commit starts to add in the notification method to notify users how much data is going to be transferred, so far only implemented for the download method. Doesn't show percentage of disk space to be taken up yet, but the basic design and implementation is how I'm thinking it'll work. If user clicks 'Confirm' it works like normal, when cancel is clicked a toastr warning message gets presented in the screen to state that the action was cancelled.

**Comment by djw8605** [comment]
Great!  Could you post a screen shot?

**Comment by robogen** [comment]
Apologies, not sure why I missed your resonse! Newest commit replicates what I did for download for upload.

While it figures out how much space it will take, the screen appears like this
hccgo_disk_think

Once it figures it out, the inner window changes to something like
hccgo_disk_found
And the data value changes depending on if the overall size can be reduced to a human readable number, showing GB if its large enough, then MB, then KB if its small enough.

If cancel is clicked the screen shows this notification
hccgo_disk_cancelled

If confirm is clicked, then it works like normal.

**Comment by djw8605** [comment]
Looks great!

-Derek

**Comment by robogen** [comment]
Thanks! Next goal is to show what percentage of the remaining disk space will be taken up, and then after that figure out some method of determining if a folder/file will be overwritten and warning the user of this, displaying the size of the folder/file to be overwritten for comparison by the user.

**Comment by djw8605** [comment]
Also, you will need to work on fixing the merge conflicts from Adam's pull request that I merged.  They're in master, so you should be able to just pull/merge from there.

**Comment by robogen** [comment]
That should fix the merge conflicts, though I was testing Adams submit script method and had an error relating to the json stuff he does. Not sure how to resolve that, but I'll see about fixing it.

**Comment by robogen** [comment]
""TypeError: Cannot read property 'jobs' of undefined\n    at Scope.$scope.writeSubmissionScript (file:///tmp/.org.chromium.Chromium.vU1IuR/js/jobSubmissionCtrl.js:161:27)\n    at $parseFunctionCall (file:///tmp/.org.chromium.Chromium.vU1IuR/lib/angular/angular.js:12456:18)\n    at callback (file:///tmp/.org.chromium.Chromium.vU1IuR/lib/angular/angular.js:21682:17)\n    at Scope.$eval (file:///tmp/.org.chromium.Chromium.vU1IuR/lib/angular/angular.js:14552:28)\n    at Scope.$apply (file:///tmp/.org.chromium.Chromium.vU1IuR/lib/angular/angular.js:14651:23)\n    at HTMLButtonElement.\u003Canonymous\u003E (file:///tmp/.org.chromium.Chromium.vU1IuR/lib/angular/angular.js:21687:23)\n    at HTMLButtonElement.jQuery.event.dispatch (file:///tmp/.org.chromium.Chromium.vU1IuR/lib/jquery/jquery.js:4435:9)\n    at HTMLButtonElement.elemData.handle (file:///tmp/.org.chromium.Chromium.vU1IuR/lib/jquery/jquery.js:4121:28)"", source: file:///tmp/.org.chromium.Chromium.vU1IuR/lib/angular/angular.js (11707)


Thats the error I'm seeing.

**Comment by djw8605** [comment]
Odd, I’m not having that problem.

Also, you will have to merge again, I just did a refresh job’s pull request: https://github.com/unlhcc/HCCGo/pull/28

-Derek

**Comment by robogen** [comment]
Using the process.env.HOME got the job submission script working for linux, and grabbed the stuff for the latest merge and tested it, everything seems to be working fine.

**Comment by robogen** [comment]
hccgo_diskavail_prototype
Playing around with how it could look for informing user how much disk space will get used up, this is what I have it looking like right now. Feels a bit wordy, what do you think?

**Comment by djw8605** [comment]
I'm always a big fan of structure.

1. Make the buttons side by side, rather than above each other.  Also, color the buttons different,cancel = red, and the other green.
2. Can you structure it so that it's more like a list:

   ```
   Download location: <location>
   Download Size: <size> GB
   Percent of available space taken: x%
   ```
3. Where is this popup going?  Is it a modal?

**Comment by robogen** [comment]
hccgo_disk_popup

How's this look? For where the popup is going, I'm just having it appear in the same screen, it's part of the page context, not it's own modal.

**Comment by djw8605** [comment]
There should be less space between the lines.

Also, add some space between the buttons.

-Derek

**Comment by robogen** [comment]
hccgo_disk_popup

Little more along these lines?

**Comment by djw8605** [comment]
Yeah, I like it.

Make the confirm on the right, and the cancel on the left.

-Derek

**Comment by robogen** [comment]
Alright, have the modification made
hccgo_disk_popup

I like the look of it. I also modified GenericClusterInterface.js slightly. Since users don't have a specific quota on a user-by-user basis, reporting on that basis doesn't provide a lot of information. I modified it so it reports based on the users group, and how much space the group itself has remaining. Landing page shows this now:
hccgo_storage

**Comment by djw8605** [comment]
Looks good.  What's left?  Upload and download work, right?

**Comment by robogen** [comment]
Yep yep, the system can upload and download properly. The only thing I can think of is currently the system just reports how much space will get used, but if the files being downloaded/uploaded would take more space than whats available the user could still engage the action.

I'd like to get it working so in that sort of a situation the user can't engage the transfer, but that's not necessarily something that 'needs' to be finished immediately, as everything is currently in working condition. If anything, we could consider this pull request finished, and I can work on that feature in a new pull.

**Comment by djw8605** [comment]
Yup, sounds good.  Go go gadget merge!

-Derek

**Comment by robogen** [comment]
Woohoo!

**Comment by robogen** [comment]
Just as one final update I went through and tested everything, submitted jobs and transferred files, everything seems to be working correctly.
